Update lesson (#2)
learn-co-curriculum/phase-1-scope-chain-readme: README.md

## Learning Goals

> - Examine how scopes can be nested.

Reviewer: Consider updating "_Examine_" to something like "_Determine_ (or _Identify_) how scopes can be nested" to make this goal even more specific and measurable.

Reviewer 2: I like "Identify" as an option!

- Demonstrate how the environment in which a function is created gets added to
  its scope chain.
- Describe how the scope chain is traversed to resolve variable lookup.

## Introduction

Every variable, every function defined in JavaScript belongs to a scope -
whether it's global, functional, or block. Previously, we've learned how to

> deduct what scope a variable belongs to. As programmers, when we read or write

Reviewer 3: ```suggestion
deduce what scope a variable belongs to. As programmers, when we read or write
```

> code, we can tell what scope a variable belongs to by just looking at its

Reviewer 2: ```suggestion
code, we can tell what scope a variable belongs to by looking at its
```

> surroundings. If it's inside a function, it's scoped to that function, and so

Reviewer 2: ```suggestion
surroundings. If it's inside a function, it's scoped to that function; if it's inside a block, it's scoped to that block; and so
```

on.

> Subsequently, we can tell what parts of our code have access to any given

Reviewer 3: ```suggestion
Similarly, we can tell what parts of our code have access to any given
```

variable. If it's inside a function, only anything _inside_ that function has
access to it. If it's at the top level of a file, it's globally scoped,

> therefore any part of our code, after the variable has been defined, has access

Reviewer 2: ```suggestion
which means that any part of our code that comes after the variable has been defined has access
```

to it.

> Again, we know this by just looking at everything around it. So, how does a

Reviewer 2: ```suggestion
As programmers, we can analyze scope just by looking at our code. But how does a
```

computer do it? When a JavaScript file runs, it has access to something called

> the **scope chain**. This chain provides all the scopes currently accessible by

Reviewer 3: Maybe "determines" rather than "provides"?

a given section of code, such as a function, or an `if` statement.

In this lesson, we'll discuss how the scope chain allows us to access variables
and functions declared in outer scopes within an inner function.

## Nested Scopes

Before we further examine the scope chain, there's an elephant in the room we
must discuss. **Scopes can be nested**.

> This should not sound surprising. After all, we have already seen examples of

Reviewer 2: ```suggestion
We have already seen examples of
```

> nested scopes. You've even written code with nested scopes. The levels of scopes

Reviewer 2: ```suggestion
nested scopes. You've even written code with nested scopes! The levels of scopes
```

> themselves demonstrate how they can be nested.

Reviewer 2: This wording is a little vague "The levels of scopes themselves demonstrate how they can be nested". Is this referring to global scope, function scope, and block scope? If so, we could change this to something like "The fact that we can have function and block scope within our global scope already demonstrates how scopes can be nested.

There is always a global scope, and any function or block statement you write

> within it will always create a nested one. For example:

Reviewer 3: ```suggestion
within it will always create a nested scope. For example:
```

```js
const outside = "Hello, I'm in the global scope!";

function foo() {
  const inside = "Hello, I'm in the function scope!";
}
```

In the above code, there are two scopes: the global, and the `foo` function
scope. The `foo` function itself belongs to the global scope, just like the
`outside` variable. However, it also creates its own scope within its function
body, where the `inside` variable belongs. Let's look at it as a bulletpoint
list:

- Global Scope
  - `outside`
  - `foo()`
    - `foo()` scope

>       - `inside`

Reviewer 3: I think graphics would be clearer than bulleted lists - although of course graphics are less accessible...

> Here, we can see how the `foo()` function scope is nested within the global one.

Reviewer 2: ```suggestion
Here, we can see how the `foo()` function scope is nested within the global scope.
```

    return totalStrikes;
  }

  if (strikes >= availableStrikes) {
    atBat = false;
    const message = player + " is out!";
    console.log(message);
  } else {
    return addStrike();
  }
}

isOut(1);
```

> <details><summary>Bonus: Before we step through the code or run it, can you determine what the above code will output?</summary><p>
> Ohtani has 2 strike(s)!

</p></details>

Breaking down the above example into our bulletpoint scope chain, we get:

- Global scope
  - `player`
  - `isOut()`
    - `strikes`
    - `atBat`
    - `availableStrikes`
    - `addStrike()`
      - `totalStrikes`
    - `if` block
      - `message`
    - `else` block

Let's walk through the code line by line, scope by scope, starting from the top.

First, we have two things in our global scope, the outermost scope. The `player`
variable, and the `isOut` function.

```js
const player = "Ohtani";

function isOut(strikes) {
  ...
}

isOut(2);
```

At the bottom of the code, `isOut()` is invoked. When the code executes and gets
to that line, it will look for that function within the current scope. The
current scope, where the function is trying to be accessed, is global. Because

> `isOut()` is a globally scoped function, it is accesible and will run

Reviewer 3: ```suggestion
`isOut()` is a globally scoped function, it is accessible and will run
```

`player` variable is used, but it does not exist within `addStrike()`. When
JavaScript looks for it in the closest outer scope, it cannot find it in
`isOut()` either. So, it goes one step further out to the global scope, where it
successfully finds `player`'s value.

**When looking up variables, JavaScript will always look at its current scope
first. Then, if unfound, it will continue moving up the scope chain until the
the variable's value is found.**

Next, let's take a look at the `if` statement block.

```js
if (strikes >= availableStrikes) {
  atBat = false;
  const message = player + " is out!";
  console.log(message);
}
```

For the conditional, it uses variables from the `isOut()` scope that it has
access to. It also uses a variable, `player`, from the global scope when
declaring `message`. In all these cases, the process is as we just discussed. It
will keep going up the scope chain until it finds the variables needed.

The new point we want to pay attention to here is the value reassignment of the
`atBat` variable.

As we previously learned, when _declaring_ a variable with no keyword, it will
become a globally scoped variable, regardless of where it's defined. In this

> case, we intuitively know, from what we've learned about JavaScript so far, that

Reviewer 3: I recommend modifying the second sentence to just: "However, that isn't what's happening here."

fuzzy, that is OK. Your sense for scope will develop the more code you write,
and the more `ReferenceError` bugs you run into and fix.

As a matter of fact, let's do some of that together.

## Trying to Break the Scope Chain

To further understand the scope chain, we must try to break it. That will let us
see how much wiggle room the scope chain has, if any.

What would happen if we try to invoke `addStrike()` _outside_ of `isOut()`?

```js
const player = "Ohtani";

function isOut(strikes) {
  let atBat = true;
  const availableStrikes = 3;

  function addStrike() {
    const totalStrikes = strikes + 1;
    console.log(player + " has " + totalStrikes + " strike(s)!");

    return totalStrikes;
  }

  if (strikes >= availableStrikes) {
    atBat = false;
    const message = player + " is out!";
    console.log(message);
  } else {
    return addStrike();
  }
}

addStrike(); // <= New addition
isOut(1);
```

It throws a `ReferenceError`, telling us that `addStrike` does not exist. This
is because the **scope chain can only go in one direction**. The chain can only
be traversed starting from inwards and going outwards. It cannot be traveled
from the outside in.

> In other words, **inner scopes _can_ access anything in its outer scopes. But

Reviewer 3: ```suggestion
In other words, **inner scopes _can_ access anything in their outer scopes. But
```

> outer scopes _cannot_ access anything in its inner scopes.**

Reviewer 3: ```suggestion
outer scopes _cannot_ access anything in their inner scopes.**
```

> Let's try one more wacky thing. What if we try to access `totalStrikes` within

Reviewer: I really don't think this lesson is too long. It all _seems_ relevant to me even though I'm no expert. If you do need to take something out though, you could consider taking out one of these walk through examples unless they are both necessary.
